web/bb-datepage.cgi.1
.TH BB-DATEPAGE.CGI 1 "Version 4.2.2: 15 Dec 2008" "Xymon"
.SH NAME
bb-datepage.cgi \- Xymon CGI script to view pre-built reports by date
.SH SYNOPSIS
.B "bb-datepage.cgi?type={day,week,month} --url=URLPREFIX [options]

.SH DESCRIPTION
\fBbb-datepage.cgi\fR is invoked as a CGI script via the bb-datepage.sh CGI wrapper.

bb-datepage.cgi is passed a QUERY_STRING environment variable with the
type of time-selection that is desired: Either "day", "week" or "month"
can be requested. It will then generate a web form with appropriate
day/week/month selection boxes, and based on the users' selection a
resulting url is built from the URLPREFIX and the time selection.
The browser is then redirected to this URL.

The URL is constructed from the URLPREFIX, the type-parameter, the 
value of the "pagepath" or "host" cookie, and the users' selection 
as follows:

.IP type=day
The final URL is \fBURLPREFIX/daily/YEAR/MONTH/DAY/PAGEPATH\fR.

.IP type=week
The final URL is \fBURLPREFIX/weekly/YEAR/WEEK/PAGEPATH\fR.

.IP type=month
The final URL is \fBURLPREFIX/monthly/YEAR/MONTH/PAGEPATH\fR.

YEAR is the full year (4 digits, including century). MONTH is the
two-digit number of the month (01..12). DAY is the number of the 
day in the month (01..31). WEEK is the ISO 8601:1988 week-number 
(01..53). PAGEPATH is the current value of the "pagepath" cookie
if set; if it is not set but the "host" cookie is set, then this
host is looked up in the bb-hosts file and the page where this host
is found is used for PAGEPATH. These two cookies are set by the 
default web-header templates supplied with Xymon.

.SH OPTIONS
.IP "--url=URLPREFIX"
This specifies the initial part of the final URL. This option is required.

.IP "--hffile=FILENAME"
Specifies the template files (from $BBHOME/web/) to use. The default is 
"--hffile=report".

.IP "--color=COLOR"
Sets the background color of the generated webpage. The default is blue.

.IP "--env=FILENAME"
Loads the environment defined in FILENAME before executing the CGI script.

.IP "--debug"
Enables debugging output.

.IP "$BBHOME/web/report_form_daily"
HTML form template for the date selection form when type=daily.

.IP "$BBHOME/web/report_form_weekly"
HTML form template for the date selection form when type=weekly.

.IP "$BBHOME/web/report_form_monthly"
HTML form template for the date selection form when type=monthly.

.IP "$BBHOME/web/report_header"
HTML header file for the generated web page

.IP "$BBHOME/web/report_footer"
HTML footer file for the generated web page

.SH "ENVIRONMENT VARIABLES"
.IP BBHOME
Used to locate the template files for the generated web pages.

.IP QUERY_STRING
Contains the parameters for the CGI script.

.SH "SEE ALSO"
bbgen(1), bb-hosts(5), hobbitserver.cfg(5)



Changes
Changes from 4.2.0 -> 4.2.2 (01 Dec 2008)
=========================================

Changelog extracted from Subversion.


------------------------------------------------------------------------
r5935 | storner | 2008-11-26 12:57:11 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   A /branches/4.2.0 (from /trunk:5040)

Create branch for maintenance of the 4.2.x release.


------------------------------------------------------------------------
r5936 | storner | 2008-11-26 13:13:50 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/common/bb-hosts.5

The ability to set a DOWNTIME for an individual test was mentioned in the Changes file, but not documented in the bb-hosts man-page.


------------------------------------------------------------------------
r5937 | storner | 2008-11-26 13:14:48 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/hobbitd/hobbitd_client.c

The hobbitd_client program would crash when running in test-mode, i.e. launched with the --test option to test disk-, procs- and ports-settings. This only applies to running it in test-mode, normal operation is not affected.


------------------------------------------------------------------------
r5938 | storner | 2008-11-26 13:15:40 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/lib/headfoot.c
   M /branches/4.2.0/web/bb-datepage.c
   M /branches/4.2.0/web/bb-eventlog.c
   M /branches/4.2.0/web/bb-findhost.c
   M /branches/4.2.0/web/bb-rep.c
   M /branches/4.2.0/web/bb-snapshot.c
   M /branches/4.2.0/web/hobbit-enadis.c
   M /branches/4.2.0/web/hobbit-nkedit.c

Some of the web page would generate an extra line with "Content-type: text/html" at the top of the webpage, although this would most often be hidden by the menubar.


------------------------------------------------------------------------
r5939 | storner | 2008-11-26 13:16:21 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/client/runclient.sh

Using extra options with the client-side "runclient.sh" script when performing a restart command fails to pass the extra options to the final command. 


------------------------------------------------------------------------
r5940 | storner | 2008-11-26 13:17:12 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/hobbitd/hobbitfetch.c

hobbitfetch insisted that hosts must have a valid IP-adress in bb-hosts. This causes problems for hosts with dynamic IP-adresses. This patch causes hobbitfetch to lookup the IP-address at run-time if the IP is listed as "0.0.0.0".


------------------------------------------------------------------------
r5941 | storner | 2008-11-26 13:20:44 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/web/hobbit-confreport.c
   M /branches/4.2.0/web/hobbitsvc-info.c

Hostname filtering on the info column page is broken, if you have multiple hosts with similar names, resulting in the same columns showing up multiple times in the disable section. A similar problem affects host filtering in the configuration report tool.


------------------------------------------------------------------------
r5942 | storner | 2008-11-26 13:21:27 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/client/hobbitclient-sunos.sh

The Solaris client would not collect the "top" process-listing data which is normally included in the data on the "cpu" column.


------------------------------------------------------------------------
r5943 | storner | 2008-11-26 13:25:28 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/build/bb-commands.sh
   M /branches/4.2.0/client/hobbitclient-sunos.sh

The Solaris lofs (loopback filesystem) should not be included in the df output appearing on the Disk status.
------------------------------------------------------------------------
r5944 | storner | 2008-11-26 13:29:56 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/lib/cgi.c
   M /branches/4.2.0/lib/cgi.h
   M /branches/4.2.0/web/bb-ack.c
   M /branches/4.2.0/web/bb-datepage.c
   M /branches/4.2.0/web/hobbit-confreport.c
   M /branches/4.2.0/web/hobbit-enadis.c
   M /branches/4.2.0/web/hobbit-hostgraphs.c
   M /branches/4.2.0/web/hobbit-statusreport.c

Cookie handling does not always work. This results in hosts not showing up on the new acknowledgment page, or hosts missing from the Metrics report.
------------------------------------------------------------------------
r5945 | storner | 2008-11-26 13:30:31 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/bbdisplay/bbgen.h

Hobbit 4.2.0 fails to build on some OS X releases due to a naming conflict between a system-defined datatype, and a different datatype defined in Hobbit.
------------------------------------------------------------------------
r5946 | storner | 2008-11-26 13:31:41 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/hobbitd/webfiles/zoom.js
   M /branches/4.2.0/lib/hobbitrrd.c
   M /branches/4.2.0/lib/hobbitrrd.h
   M /branches/4.2.0/lib/htmllog.c
   M /branches/4.2.0/web/hobbitgraph.c
   M /branches/4.2.0/web/hobbitsvc-trends.c

When viewing detailed graphs linked from the detailed status page, the background color of the webpages would always show green, regardless of the actual color of the status it was linked from. This patch keeps the background color at what the status page shows.
------------------------------------------------------------------------
r5947 | storner | 2008-11-26 13:32:55 +0100 (Wed, 26 Nov 2008) | 1 line
Changed paths:
   M /branches/4.2.0/hobbitd/webfiles/confreport_front
   M /branches/4.2.0/hobbitd/wwwfiles/menu/menu_items.js.DIST
   M /branches/4.2.0/web/Makefile
   M /branches/4.2.0/web/hobbit-confreport.c
   M /branches/4.2.0/web/hobbit-confreport.cgi.1

The Configuration Report did not include information about what systems appears on the "Critical Systems" view.

web/bb-datepage.c
/*----------------------------------------------------------------------------*/
/* Hobbit webpage generator tool.                                             */
/*                                                                            */
/* Copyright (C) 2004-2006 Henrik Storner <henrik@storner.dk>                 */
/*                                                                            */
/* This program is released under the GNU General Public License (GPL),       */
/* version 2. See the file "COPYING" for details.                             */
/*                                                                            */
/*----------------------------------------------------------------------------*/

static char rcsid[] = "$Id: bb-datepage.c,v 1.15 2006-07-11 17:18:22 henrik Exp $";

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <fcntl.h>
#include <limits.h>

#include "libbbgen.h"
#include "version.h"

static enum { FRM_NONE, FRM_MONTH, FRM_WEEK, FRM_DAY } frmtype = FRM_NONE;
static int year = -1;
static int month = -1;
static int day = -1;
static int week = -1;

static void errormsg(char *msg)
{
	printf("Content-type: %s\n\n", xgetenv("HTMLCONTENTTYPE"));
	printf("<html><head><title>Invalid request</title></head>\n");
	printf("<body>%s</body></html>\n", msg);
	exit(1);
}

static void parse_query(void)
{
	cgidata_t *cgidata = cgi_request();
	cgidata_t *cwalk;

	if (cgidata == NULL) {
		errormsg(cgi_error());
	}

	cwalk = cgidata;
	while (cwalk) {
		if (strcasecmp(cwalk->name, "YEAR") == 0) {
			year = atoi(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "MONTH") == 0) {
			month = atoi(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "DAY") == 0) {
			day = atoi(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "WEEK") == 0) {
			week = atoi(cwalk->value);
		}
		else if (strcasecmp(cwalk->name, "TYPE") == 0) {
			if (strcasecmp(cwalk->value, "month") == 0) frmtype = FRM_MONTH;
			else if (strcasecmp(cwalk->value, "week") == 0) frmtype = FRM_WEEK;
			else if (strcasecmp(cwalk->value, "day") == 0) frmtype = FRM_DAY;
			else errormsg("Bad type parameter\n");
		}

		cwalk = cwalk->next;
	}
}

int main(int argc, char *argv[])
{
	int argi;
	char *hffile = "report";
	char *urlprefix = "";
	int bgcolor = COL_BLUE;
	char *envarea = NULL;

	for (argi = 1; (argi < argc); argi++) {
		if (argnmatch(argv[argi], "--env=")) {
			char *p = strchr(argv[argi], '=');
			loadenv(p+1, envarea);
		}
		else if (argnmatch(argv[argi], "--area=")) {
			char *p = strchr(argv[argi], '=');
			envarea = strdup(p+1);
		}
		else if (strcmp(argv[argi], "--debug") == 0) {
			debug = 1;
		}
		else if (argnmatch(argv[argi], "--hffile=")) {
			char *p = strchr(argv[argi], '=');
			hffile = strdup(p+1);
		}
		else if (argnmatch(argv[argi], "--color=")) {
			char *p = strchr(argv[argi], '=');
			bgcolor = parse_color(p+1);
		}
		else if (argnmatch(argv[argi], "--url=")) {
			char *p = strchr(argv[argi], '=');
			urlprefix = strdup(p+1);
		}
	}

	redirect_cgilog("bb-datepage");
	parse_query();

	if (cgi_method == CGI_POST) {
		char *pagepath, *cookie, *endurl;

		cookie = get_cookie("pagepath");

		if (cookie && *cookie) {
			pagepath = strdup(cookie);
		}
		else {
			cookie = get_cookie("host");

			if (cookie && *cookie) {
				namelist_t *hinfo;

				load_hostnames(xgetenv("BBHOSTS"), NULL, get_fqdn());
				hinfo = hostinfo(cookie);
				if (hinfo) {
					pagepath = bbh_item(hinfo, BBH_PAGEPATH);
				}
				else {
					pagepath = "";
				}
			}
			else {
				pagepath = "";
			}
		}

		endurl = (char *)malloc(strlen(urlprefix) + strlen(pagepath) + 1024);

		switch (frmtype) {
		  case FRM_DAY:
			if ((year == -1) || (month == -1) || (day == -1)) errormsg("Invalid day-request");
			sprintf(endurl, "%s/daily/%d/%02d/%02d/%s", urlprefix, year, month, day, pagepath);
			break;

		  case FRM_WEEK:
			if ((year == -1) || (week == -1)) errormsg("Invalid week-request");
			sprintf(endurl, "%s/weekly/%d/%02d/%s", urlprefix, year, week, pagepath);
			break;

		  case FRM_MONTH:
			if ((year == -1) || (month == -1)) errormsg("Invalid month-request");
			sprintf(endurl, "%s/monthly/%d/%02d/%s", urlprefix, year, month, pagepath);
			break;

		  case FRM_NONE:
			break;
		}

		if (*pagepath) strcat(endurl, "/");

		fprintf(stdout, "Content-type: %s\n\n", xgetenv("HTMLCONTENTTYPE"));
		fprintf(stdout, "<html><head><meta http-equiv=\"refresh\" content=\"0; URL=%s\"></head></html>\n", endurl);
	}
	else if (cgi_method == CGI_GET) {
                char formfn[PATH_MAX];
		time_t seltime;
		struct tm *seltm;

		seltime = time(NULL); seltm = localtime(&seltime);

                /* Present the query form */
		switch (frmtype) {
		  case FRM_DAY:
			seltm->tm_mday -= 1; seltime = mktime(seltm);
			sprintf(formfn, "%s_form_daily", hffile);
			break;

		  case FRM_WEEK:
			seltm->tm_mday -= 7; seltime = mktime(seltm);
			sprintf(formfn, "%s_form_weekly", hffile);
			break;

		  case FRM_MONTH:
			seltm->tm_mon -= 1; seltime = mktime(seltm);
			sprintf(formfn, "%s_form_monthly", hffile);
			break;

		  case FRM_NONE:
			errormsg("No report type defined");
		}

		sethostenv("", "", "", colorname(bgcolor), NULL);
		fprintf(stdout, "Content-type: %s\n\n", xgetenv("HTMLCONTENTTYPE"));
		showform(stdout, hffile, formfn, COL_BLUE, seltime, NULL, NULL);
	}

	return 0;
}

